Proposal: convert f090 Mining Reserve actor to a keyless account actor (#903)
filecoin-project/FIPs: FIPS/fip-keyless-f090-account.md

---
fip: "TBD" 
title: Convert f090 Mining Reserve actor to a keyless account actor
author: Jennifer Wang (@jennijuju), Jon Victor (@jnthnvctr)
discussions-to: https://github.com/filecoin-project/FIPs/discussions/901
status: Draft
type: Technical
category: Core
created: 2024-01-04
---

# Convert f090 Mining Reserve actor to a keyless account actor

## Simple Summary

Covert f090 (FIL mining reserve) actor type from a _multisig_ to an _account_ that is keyless (like `f099`).

## Abstract

> At the moment, the actor holding the FIL mining reserve is `f090`, which is a _multisig_ actor with 3 signers and an approval threshold of 2. All three signers are _account_ actors, with wallet keys to those accounts. Such actor type leaves the ownership and control over the mining reserve funds to signers of the _multisig_ actor, which opposes the decentralized governance principle over network funds and creates potential security leaks. This FIP proposes to Covert f090 (FIL mining reserve) actor type from a multisig to an account, and further transfer the ownership, control and governance of the funding reserved from the signers of the multisig back to the network's participants via community governance.

Reviewer: Some editorial changes for clarity.

```suggestion
At the moment, the actor holding the FIL mining reserve is a _multisig_ actor sitting at address `f090`. It has 3 signers and an approval threshold of 2. All three signers are _secp256k1 account_ actors, secured by corresponding wallet keys. This setup leaves the ownership and control over the mining reserve funds to signers of the _multisig_ actor, which undermines the decentralized governance principle over network funds, and creates potential security leaks.

This FIP proposes to alter the type of the FIL mining reserve actor from the aforementioned multisig to a keyless account actor, and further transfer the ownership, control and governance of the funding reserved from the signers of the multisig back to the network's participants via community governance.
```

## Change Motivation

As documented in the spec [here](https://spec.filecoin.io/#section-systems.filecoin_token.token_allocation), upon network launch, a pool of tokens was allocated as a Mining Reserve to support the maintenance of a robust Filecoin economy. It also mentions, "It will be up to the community to determine in the future how to distribute those tokens, through Filecoin improvement proposals (FIPs) or similar decentralized decision-making processes." It is worth noting that when the spec was written and when the network was first launched, the FIP process was not yet well-designed or matured to support network governance, and significant improvements have been made since then.

> At the moment, the actor holding the FIL mining reserve is `f090`, which is a multisig actor with 3 signers and an approval threshold of 2. All three signers are account actors, with wallet keys to those accounts. This means three individuals hold the keys to three wallets and can move funds and change the terms of f090, as long as 2 out of 3 signers approve the same transactions. They can do so without anyone else's approval, protocol-wise speaking. This opposes the decentralized governance principle as quoted in the above paragraph from the spec. This also creates a security black hole in the network where if 2 out of the 3 signer keys of f090 are compromised by malicious actors, they may cause serious economic damage to the Filecoin network.

Reviewer: ```suggestion
At the moment, the actor holding the FIL mining reserve is `f090`, which is a multisig actor with 3 signers and an approval threshold of 2. All three signers are account actors, with wallet keys to those accounts. This means three individuals hold the keys to three wallets and can move funds and change the terms of f090, as long as 2 out of 3 signers approve the same transactions. They can do so without anyone else's approval, protocol-wise speaking. This opposes the decentralized governance principle as quoted in the above paragraph from the spec. This also creates a security gap in the network where if 2 out of the 3 signer keys of f090 are compromised by malicious actors, they may cause serious economic and reputational damage to the Filecoin network.
```

> In addition, having the reserve holds in a multisig also means changes toward the mining reserve, including unlock portion of the funds to new network participants or any operation/mechanism accepted by the community via FIP, may need to be managed by the msig signers via signed transactions which is a huge operational overhead (and again against the principle).

Reviewer: Would consider rephrasing this. I think the point you're trying to emphasise is that there's an assumption that these multisig operators would abide by the community's decision, would act in honesty, would custody the keys securely, and would remain readily available to sign the corresponding messages when determined by the community.

Reviewer 2: A network upgrade could always override absent/lost signers. That's a lot more operational overhead than msig. I'm not sure this paragraph is saying something useful (though what Raul suggests as an alternative would be, in terms of the additional assumptions everyone has to make).

Reviewer 3: ```suggestion
In addition, having the reserve held in a multisig also means changes regarding use the reserve may need to be managed by the msig signers via signed transactions which is a huge operational overhead (and again against the principle).
```

I think if you're going to avoid any "intended use of the funds" as per point @momack2 raised below then you just need to avoid usage suggestions all together.

The proposal here is rather simple: to convert the f090 actor type from a *multisig* to an *account* type (without any keys, like `f099`), and any changes to the mining reserve must be proposed via FIPs, governed by the FIP and network upgrade process.

## Specification

> Convert f090 actor type from a _multisig_ to _account_ type. The new account actor should be keyless, like `f099`.

Reviewer: The state schema of the Account actor is:

```rust
pub struct State {
    pub address: Address,
}
```

The burnt funds contains a self-reference to its ID address in the `address` field (address = f099). I assume this proposal wants to use the same setup, so we should specify it here. (That's what gives it the "keyless" property).

## Specification

Convert f090 actor type from a _multisig_ to _account_ type. The new account actor should be keyless, like `f099`.

## Design Rationale

The chose of making f090 actor to be a keyless account type is to ensure that no single actor/entity owns the control over Filecoin mining reserves. The use of the mining reserves will be proposed and governed by the FIP process, and further adopted by the network via network upgrades. 

## Backwards Compatibility

This change is not backwards compatible, upon the activation of this FIP in network upgrade:
- the actor type of f090 will be migrated from a _multisig_ to _account_.
- the existing signers of f090 multisig will no longer be able to perform multisig operations of the actor

## Test Cases

- Get actor cid of f090 it should return actor code cid of an `account`

## Security Considerations

This proposal improves the network security by removing the ownership and control of 15% of the total network token supply from 3 individuals to the network participants.

## Product Considerations

> This proposal allows the governance of the network reserved funds to be in hands of the network participants rather than 3 key owners, which makes the network more decentralized. 

Reviewer 4: I do think more should be written about the **intended use of these funds** (as previously laid out at ICO in 2017 and prior to network launch in 2020), and ensure that any technical change here is well considered and inline with the best product interests of Filecoin. A non-goal (IMHO) would be encouraging a deluge of un-useful FIPs from many authors aiming to self-allocate resources from this capital pool that thereby inflate circulating supply without achieving the intended goal (retrieval mining IIRC).

Author: > I do think more should be written about the intended use of these funds

This FIP is not proposing any change towards the intended use of the funds, therefore i don't have much more to add other than linking the spec. 

> . A non-goal (IMHO) would be encouraging a deluge of un-useful FIPs from many authors aiming to self-allocate resources from this capital pool that thereby inflate circulating supply without achieving the intended goal 

I agree with the non-goal, and I think it is orthogonal with this FIP proposal. The mega point of this FIP is that any protocol change, including the use of the reserved network funds should be governed by the FIP and network upgrade process. And the FIP process should prevent the funds in the mining reserve to be moved without achieving its intended goal (whats the intended goal should also be a part of the FIP discussion when ideas coming up imho). With f090 being a multisig, reserved funds could be used/moved against the intended goal if the key holders of the msig signer decided to act maliciously, i.e: they can burned the funds by approve a tx to send $$ of fil from f090 to f099; they might draining the funds for personal interest and so on.

Author: accepted this suggestion https://github.com/filecoin-project/FIPs/pull/903#discussion_r1448191658

> This proposal allows the governance of the network reserved funds to be in hands of the network participants rather than 3 key owners, which makes the network more decentralized. 

Reviewer 2: "More decentralised" is always a bit handwavy. But I think you could point to the product benefits of much higher relative certainty over the (non-)disbursement of those funds without significant community advertisement, consultation, and agreement.

Reviewer 3: ```suggestion
This proposal places governance of the reserved funds entirely in hands of the network participants rather than 3 key owners, thus providing greater transparency and assurance of community consultation and deliberation in appropriate future disbursement of the funds.
```

to repeat a point already made repeatedly: "allows the governance ..." - it's always been allowed, a FIP could just take funds without needing a signer, leaving it a msig. All this is doing is make it absolutely clear that FIP is the only path to doing it now.

This FIP proposes to alter the type of the FIL mining reserve actor from the aforementioned multisig to a keyless account actor, and further transfer the ownership, control and governance of the funds reserved from the signers of the multisig back to the network's participants via community governance.

## Change Motivation

As documented in the spec [here](https://spec.filecoin.io/#section-systems.filecoin_token.token_allocation), upon network launch, a pool of tokens was allocated as a Mining Reserve to support the maintenance of a robust Filecoin economy. It also mentions, "It will be up to the community to determine in the future how to distribute those tokens, through Filecoin improvement proposals (FIPs) or similar decentralized decision-making processes." It is worth noting that when the spec was written and when the network was first launched, the FIP process was not yet well-designed nor matured to support network governance, and significant improvements have been made since then.

> At the moment, the actor holding the FIL mining reserve is `f090`, which is a multisig actor with 3 signers and an approval threshold of 2. All three signers are account actors, with wallet keys to those accounts. This means three individuals hold the keys to three wallets and can move funds and change the terms of f090, as long as 2 out of 3 signers approve the same transactions. They can do so without anyone else's approval, speaking protocol-wise. This opposes the decentralized governance principle as quoted in the above paragraph from the spec. This also creates a security gap in the network where if 2 out of the 3 signer keys of f090 are compromised by malicious actors, they may cause serious economic and reputational damage to the Filecoin network.

Reviewer 5: ```suggestion
At the moment, the actor holding the FIL mining reserve is `f090`, which is a multisig actor with 3 signers and an approval threshold of 2. All three signers are account actors, with wallet keys to those accounts. This means that 2 out of 3 _individuals_ hold the keys to move funds and change the terms of f090. They can do so without a governance process or any protocol-level safeguards, which conflicts with the decentralized governance principle outlined in the spec. It further creates security concerns for the whole Filecoin network, in that a malicious actor that compromises 2 out of the 3 signer keys of f090 gains the ability to cause serious economic and reputational damage.
```

> In addition, having the reserve held in a multisig also means changes regarding use the reserve may need to be managed by the msig signers via signed transactions which is a huge operational overhead (and again against the principle).

Reviewer 5: ```suggestion
In addition, having the reserve held in a multisig also means the use of the reserve may need to be managed via signed transactions, creating significant operational overhead (and again deviating from the governance principle).
```

> The proposal here is rather simple: to convert the f090 actor type from a *multisig* to an *account* type (without any keys, like `f099`), and any changes to the mining reserve must be proposed via FIPs, governed by the FIP and network upgrade process.

Reviewer 5: ```suggestion
This proposal is simple: to convert the f090 actor type from a *multisig* to an *account* type (without any keys, like `f099`), so that any changes to the mining reserve must be proposed via the appropriate FIP and network upgrade processes.
```

## Specification

Convert f090 actor type from a _multisig_ to _account_ type. The new account actor should be keyless, like `f099`.

```rust
pub struct State {
  pub address: Address,

>   

Reviewer 3: ```suggestion
```

}
```

The new f090 account actor will contain a self-reference to its ID address in the `address` field.

## Design Rationale

The choice of making f090 actor to be a keyless account type is to ensure that no single actor/entity owns the control over Filecoin mining reserves. The use of the mining reserves will be proposed and governed by the FIP process, and further adopted by the network via network upgrades. 

## Backwards Compatibility

> This change is not backwards compatible, upon the activation of this FIP in network upgrade:
> - the actor type of f090 will be migrated from a _multisig_ to _account_.
> - the existing signers of f090 multisig will no longer be able to perform multisig operations of the actor

Reviewer 5: ```suggestion
This change is not backwards compatible. Upon the activation of this FIP in a network upgrade:
- the actor type of f090 will be migrated from _multisig_ to _account_.
- the existing signers of the f090 multisig will no longer be able to perform operations on the actor
```

## Test Cases

> - Get actor cid of f090 it should return actor code cid of an `account`

Reviewer 5: ```suggestion
- Get actor cid of f090 should return actor code cid of an `account`
```

## Security Considerations

> This proposal improves the network security by removing the ownership and control of 15% of the total network token supply from 3 individuals to the network participants.

Reviewer 5: ```suggestion
This proposal improves the network security by removing the ownership and control of 15% of the total network token supply from 3 individuals and putting it under the control of network participants via the appropriate governance processes.
```
